### ISSUE TITLE
Update: JSON_Web_Token_for_Java_Cheat_Sheet

### ISSUE THREAD

**Issue by Tib3rius** [OWASP/CheatSheetSeries#1458]
## What is missing or needs to be updated?
The [Token Storage on Client Side](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/JSON_Web_Token_for_Java_Cheat_Sheet.md#token-storage-on-client-side) page recommends using Session Storage over Local Storage, seemingly because Local Storage persists between browser restarts (accurate), however it makes no mention of the fact that Session Storage is implemented per-tab and not per-site. That is, if example.com adds some item to Session Storage, and a user then opens a new tab through any method other than "duplicate tab" (e.g. right-click -> open link in new tab), the new tab will have a completely new (and empty) Session Storage instance, even if the user goes to example.com.

Since opening things in multiple tabs is pretty common user behavior, it is not realistic (IMO) for the recommended method to be implemented by developers.

## How should this be resolved?
Honestly, I think the recommendation should be to use Local Storage but with tokens that have short expiration times. There's no perfect solution, but Session Storage doesn't work at all so it's surprising to see it being recommended.

**Comment by jmanico** [comment]
ASVS also recommends session storage over local storage and I'd like to be consistent with that.

If you want to suggest localStorage please please make sure you state this is only a suggestion if the token exp time is low. Overall I do not suggest localStorage for most services.

**Comment by Tib3rius** [comment]
@jmanico right but surely that just means ASVS needs changing? We shouldn't be recommending developers use a storage medium for session tokens that deletes them when their users do the simple act of opening a page in a new tab, IMO.

Do you know of any modern app that actually stores them in Session Storage, and how that app maintains sessions with new tabs? If you want to stick with Session Storage, then the advice at least needs to be updated to mention the downsides and how to get around them...

**Comment by jmanico** [comment]
I suggest that the USERNAME can be stored long term, but features that maintain a long term active session is discouraged for secure software. In general we suggest 15-30 minute idle timeout and a absolutely timeout of like 8 hours.

So storing a active session is localstorage that keeps a user logged in long term is discouraged for any secure software.

**Comment by Tib3rius** [comment]
@jmanico I appreciate the response but you aren't really answering any of my questions here, and I never once mentioned long term active sessions. My point from the start has been that the OWASP recommendations simply don't work practically for modern websites / browsers.

How does a secure application maintain a session, stored in Session Storage, when a user opens a link in a new tab?

It is this fact that I am most concerned about, because if developers were to implement OWASP recommendations as they currently stand, most if not all applications would break. That to me, seems like something that at the very least we should address by **mentioning** the limitations of Session Storage on this page, and by suggesting a perhaps less secure method for developers who need users to remain logged in if they open separate tabs.

**Comment by jmanico** [comment]
I am convinced you are correct. But I would only advise the use of localStorage if the token has both idle and absolute timeout. Let me leave this open for now and I'll make a comment in ASVS and cross-reference this issue. I am with you now.

**Comment by Tib3rius** [comment]
Thanks @jmanico! Appreciate the willingness to have the discussion, I agree localStorage with severe timeout restrictions seems like a good compromise. This is a problem with seemingly no "nice" solution.

**Comment by jmanico** [comment]
One more note, this is how some folks deal with the sessionStorage issue: 

_In one application I use, the session token is indeed not shared across tabs. When you open a new tab with the application, you are redirected to the SSO provider and automatically logged in into the application. So while the application's token is stored in sessionStorage, it still works with multi-tab browsing because of the interaction with the SSO provider._

**Comment by Pastekitoo** [comment]
Hi, I would like to work on this issue. Can you please assign it to me ?

**Comment by Pastekitoo** [comment]
After reading all the discussions here and [there](https://github.com/OWASP/ASVS/issues/2076#issue-2524591813), I just need a clarification. I understood that we have to provide security recommendations for apps which people expect to work in multiple tabs, even if it's less secure. So document LocalStorage, SSOProvider and their restrictions, in addition to what's already there for SessionStorage?

**Comment by mackowski** [comment]
@Pastekitoo exactly, we know it is less secure but usability is also important. If we can add a security recommendations and explain why it is less secure it will allow to make good decisions when creating the application + make it as secure as possible.
@Tib3rius @jmanico do you agree?

**Comment by jmanico** [comment]
I think your suggestions are very helpful, @mackowski
